larpandoracontent/LArThreeDReco/LArTwoViewMatching/TwoViewTransverseTracksAlgorithm.cc

```diff
@@ -36,19 +56,105 @@ void TwoViewTransverseTracksAlgorithm::CalculateOverlapResult(const Cluster *con
+    const DiscreteProbabilityVector resampledDiscreteProbabilityVector2(discreteProbabilityVector2, resamplingPointsX);
+
+    const float correlation(LArDiscreteProbabilityHelper::CalculateCorrelationCoefficient(
+        resampledDiscreteProbabilityVector1, resampledDiscreteProbabilityVector2));
+
+    const float pvalue(LArDiscreteProbabilityHelper::CalculateCorrelationCoefficientPValueFromPermutationTest(
+        resampledDiscreteProbabilityVector1, resampledDiscreteProbabilityVector2, m_randomNumberGenerator, m_nPermutations));
+
+    const float matchingScore(1.f-pvalue);
 
-    if (xOverlap > std::numeric_limits<float>::epsilon())
-        this->GetMatchingControl().GetOverlapMatrix().SetOverlapResult(pCluster1, pCluster2, xOverlap);
+    const float locallyMatchedFraction(CalculateLocalMatchingFraction(resampledDiscreteProbabilityVector1, 
+        resampledDiscreteProbabilityVector2));
+
+    overlapResult = TwoViewTransverseOverlapResult(matchingScore, resampledDiscreteProbabilityVector1.GetSize(), 
+        correlation, locallyMatchedFraction, twoViewXOverlap);
+
+    return STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+float TwoViewTransverseTracksAlgorithm::CalculateLocalMatchingFraction(const DiscreteProbabilityVector &discreteProbabilityVector1,
+    const DiscreteProbabilityVector &discreteProbabilityVector2)
+{
+    if (discreteProbabilityVector1.GetSize() != discreteProbabilityVector2.GetSize() || 
+            0 == discreteProbabilityVector1.GetSize()*discreteProbabilityVector2.GetSize())
+        throw STATUS_CODE_INVALID_PARAMETER;
+
+    if (m_minSamples > discreteProbabilityVector1.GetSize())
+        throw STATUS_CODE_INVALID_PARAMETER;
+
+    std::vector<float> localValues1, localValues2;
+    int nMatchedComparisons(0);
+
+    for (size_t iValue = 0; iValue < discreteProbabilityVector1.GetSize(); ++iValue)
+    {
+        localValues1.emplace_back(discreteProbabilityVector1.GetProbability(iValue));
+        localValues2.emplace_back(discreteProbabilityVector2.GetProbability(iValue));
+        if (localValues1.size() == m_minSamples)
+        {
+            const float localPValue(LArDiscreteProbabilityHelper::CalculateCorrelationCoefficientPValueFromPermutationTest(
+                localValues1, localValues2, m_randomNumberGenerator, m_nPermutations));
+
+            if ((1.f-localPValue) - m_localMatchingScoreThreshold > std::numeric_limits<float>::epsilon())
+                nMatchedComparisons++;
+            localValues1.erase(localValues1.begin());
+            localValues2.erase(localValues2.begin());
+        }
+    }
+
+    const unsigned int nComparisons(static_cast<unsigned int>(discreteProbabilityVector1.GetSize())-(m_minSamples-1));
+    return static_cast<float>(nMatchedComparisons)/static_cast<float>(nComparisons);
```

Comment:
Could this just be a const float to avoid the static cast on the following line as it is only used there?

Reply:
It could, but should it?
I'd coded this like this as `nComparisons` in isolation is a count and thought it made the code slightly more readable.  I figured seeing that as a float would raise questions about why it was a float.

I am not strongly wedded to that though, if you feel strongly I'll update.

Reply:
No, I think it is also fine how it is.

Reply:
Also think it's fine as is, provided that nComparisons isn't zero (or this is checked and handled as an exception), and provided that the subtraction of the two unsigned doesn't "go negative" and give you a large positive unsigned. Also need to handle this. Fiddly!

Reply:
@johnmarshall80 I think my construction those two things cannot happen.  The code, elsewhere, constrains the size of the vectors to be at least m_minSamples long.  I can add in throwing an exception though, just in case someone mis-edits upstream code.

Reply:
I think that, on balance, the nicest middle ground is to make nComparisons be an int, rather than unsigned int.  That way, a single piece of logic can dictate whether to throw or not.

Reply:
Done.